core/formatter.cpp

```diff
@@ -1724,16 +1733,233 @@ class FixIndentation {
+            idents.push_back(elem.bind->var->name);
+        }
+        return std::unique(idents.begin(), idents.end()) != idents.end();
+    }
+
+    /// Check if the import group ends after this local
+    bool groupBoundaryAfter(Local *local) {
+        Local *next = goodLocalOrNull(local->body);
+        if (!next) {
+            return true;
+        }
+
+        bool newlineReached = false;
+        for (const auto &fodderElem: open_fodder(next)) {
+            if (newlineReached || fodderElem.blanks > 0) {
+                return true;
+            }
+            if (fodderElem.kind != FodderElement::Kind::INTERSTITIAL) {
+                newlineReached = true;
+            }
+        }
+        return false;
+    }
+
+    void ensureCleanNewline(Fodder &fodder) {
+        if (fodder.size() == 0 || fodder.back().kind == FodderElement::Kind::INTERSTITIAL) {
+            fodder.push_back(FodderElement(FodderElement::Kind::LINE_END, 0, 0, {}));
+        }
+    }
+
+    AST *toplevelImport(Local *local, ImportElems &imports, const Fodder &groupOpenFodder) {
+        assert(isGoodLocal(local));
+
+        Fodder commentAfter, beforeNextFodder;
+        std::tie(commentAfter, beforeNextFodder) = splitOpenFodder(local->body);
+
+        ensureCleanNewline(commentAfter);
+
+        ImportElems newImports = extractImportElems(local, commentAfter);
+        imports.insert(imports.end(), newImports.begin(), newImports.end());
+
+        if (groupBoundaryAfter(local)) {
+            sortGroup(imports);
+
+            Fodder afterGroup = imports.back().afterFodder;
+            ensureCleanNewline(beforeNextFodder);
+            auto nextOpenFodder = concat_fodder(
+                afterGroup,
+                beforeNextFodder
+            );
+
+            // Process the code after the current group:
+            AST *bodyAfterGroup;
+            Local *next = goodLocalOrNull(local->body);
+            if (next) {
+                // Another group of imports
+                ImportElems nextImports;
+                bodyAfterGroup = toplevelImport(next, nextImports, nextOpenFodder);
+            } else {
+                // Something else
+                bodyAfterGroup = local->body;
+                open_fodder(bodyAfterGroup) = nextOpenFodder;
+            }
+
+            return buildGroupAST(imports, bodyAfterGroup, groupOpenFodder);
+        } else {
+            assert(beforeNextFodder.empty());
+            return toplevelImport(dynamic_cast<Local*>(local->body), imports, groupOpenFodder);
+        }
+    }
+
+    void file(AST *&body)
+    {
+        ImportElems imports;
+        Local *local = goodLocalOrNull(body);
+        if (local) {
+            body = toplevelImport(local, imports, open_fodder(local));
+        }
+    }
+};
 
 std::string jsonnet_fmt(AST *ast, Fodder &final_fodder, const FmtOpts &opts)
 {
     Allocator alloc;
 
     // Passes to enforce style on the AST.
+    if (opts.sortImports)
```

Comment:
Little nit, but I think here it's generally recommended to include braces for one line if-statements (I could be wrong)